### ISSUE TITLE
Question : how many epochs to get an ok result?

### ISSUE THREAD

**Issue by pfeatherstone** [zylo117/Yet-Another-EfficientDet-Pytorch#64]
How many epochs are required to get an ok result?
I'm finding it takes the age of the universe to train efficientdet0.

**Comment by pfeatherstone** [comment]
Could we improve convergence using ciou loss ?

**Comment by zylo117** [comment]
it does take ages,
and no trick but to buy more gpus

**Comment by pfeatherstone** [comment]
I've tried ciou loss. It does seem to help. But I need to tune some loss gains, i.e.
```loss = a * cls_loss + b * iou_loss```
I need to tune a and b

**Comment by zylo117** [comment]
please pull the latest code, the former loss function is bugged

**Comment by pfeatherstone** [comment]
The FocalLoss looks exactly the same

**Comment by pfeatherstone** [comment]
Just noticed. Thanks

**Comment by DecentMakeover** [comment]
@pfeatherstone my regression loss reduces far more quickly than my classification loss, how could i tune the ciou loss to balance that ? `loss = a * cls_loss + b * iou_loss`

Thanks

**Comment by pfeatherstone** [comment]
I would experiment a bit. I’ve put this work on hold for a while. I guess the regression loss converging faster isn’t necessarily a bad thing. You get good boxes faster then it can fine tune classes. I was going to try adapting the gains at run time. So have a higher ciou loss to start off with then balance it out. It’s dark art this stuff. At the end of the day it’s going to train for at least 100 epochs so you will get there in the end

**Comment by pfeatherstone** [comment]
I might be talking crap. The cls loss, which contains the objectness somewhere, is what allows you to detect boxes in the first place. So you want that to converge faster. In yolov3 for example you want the object loss and no-object loss a bit higher than the box loss to start off with. One thing I did do in this repo is switch all the loss reductions from ‘mean’ to ‘sum’.

**Comment by DecentMakeover** [comment]
'mean' to 'sum' , did it help ?

**Comment by pfeatherstone** [comment]
For starters, I switched all the loss reductions from ‘mean’ to ‘sum’. From memory, that will scale your cls loss a lot more

**Comment by pfeatherstone** [comment]
> 'mean' to 'sum' , did it help ?
Well it will scale your cls loss a lot more.

**Comment by DecentMakeover** [comment]
alright, thanks ill try that. 

for now ill skip the ciou loss, i am not able to understand it very well.
Thanks for the inputs.

**Comment by pfeatherstone** [comment]
To be honest, I got bored of messing around with it. It works and It’s got at least 100 epochs to converge. You can also mess around with optimisers and schedulers . There are 100s of ways you can spend fine tuning stuff.

**Comment by pfeatherstone** [comment]
I pinched the ciou loss code from the ultralytics yolov3 repo. It’s very good.

**Comment by DecentMakeover** [comment]
thanks ill check, i wanted to know your thoughts on number of epochs to train, is 500 epochs an overkill for a dataset of about 1200 images ?

**Comment by pfeatherstone** [comment]
The architecture of efficientdet is very deep. It has so many connections. It could be why it takes longer to train. The model may be small but training-wise there are a lot of activations held in memory. I can only use a batchsize of 8 on my GPU. That also doesn’t help. I don’t have a fancy rig.

**Comment by pfeatherstone** [comment]
In my opinion if you need 500 epochs to converge then it’s no good

**Comment by pfeatherstone** [comment]
It’s likely going to be overfitted

**Comment by DecentMakeover** [comment]
oh

**Comment by DecentMakeover** [comment]
yeah even i thought it was an overkill 100-150 seems more reasonable

**Comment by pfeatherstone** [comment]
Hence why I was messing around with other losses to help with convergence. The less epochs required the better.

**Comment by pfeatherstone** [comment]
Yep I agree. I never go past 100 epochs

**Comment by DecentMakeover** [comment]
Hmmm, alright let me see

**Comment by pfeatherstone** [comment]
You could mess around with higher learning rates using SGD. I would recommend reading the papers on super-convergence using OneCycle policy

**Comment by DecentMakeover** [comment]
sure, ill check  those out

**Comment by pfeatherstone** [comment]
i’m waiting for the day some academic spews out an optimal black box for training nets

**Comment by DecentMakeover** [comment]
ill join that club
